### PR TITLE
ci: add release check

### LINKED ISSUE #3514: Proposal: reuse test in CI for the release check
### Background:

Currently, before we publish a release, there would be a call for the manual check for the release tarball in the email list, while we better should not rely on the manual check to find problems. CI could better serve this purpose.

### Github Actions Steps

1. The action would be triggered when the PR is (1) not in master branch (2) with tag `release-test`
2. Upload the package tarball to the PR, since the test step would be before the release so we need to do that ourselves. The file would be less than 200k for now, so it is tolerable for GitHub.
3. Get the file with `git ls-files -m` since that's the only new file. Unzip it.
4. Add test files and all other essential gadgets to it.
5. Do the test, like the normal CI

### PR DESCRIPTION
**Pull request by Yiyiyimu**
Signed-off-by: yiyiyimu <wosoyoung@gmail.com>

### What this PR does / why we need it:
fix #3514

So the CI basically do:
1. Whenever push to branch start with 'v' (for us it's always release branch), release part would be triggered.
2. Make release, remove everything except tarball, `.travis` and `t`. And untar for test
3. Reuse CI in ubuntu for test
4. Upload tarball as an artifact, so we could directly download it in the workflow page [Ref](https://docs.github.com/en/actions/managing-workflow-runs/downloading-workflow-artifacts)

Some problems need suggestions:
~1. Github actions provide [composite](https://docs.github.com/en/actions/creating-actions/creating-a-composite-run-steps-action) run so we could reuse code in other actions. But since it does not support matrix yet, so we could not use it in this situation. I just copied `build.yml` and it's not so satisfying~
2. Using matrix, one tarball would be created and uploaded for each environment. It's not a big deal since they are all the same, but still not satisfying.

**TODO**:
- [ ] ~need to upload someone's secret key, so we could use `gpg`~
- [x] Currently for test, it would run on pull request. It should be changed to run on push before merge

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: .github/workflows/release-test.yml
Comment (spacewander) [inline .github/workflows/release-test.yml:19]: Seems we only need the linux_openresty_xxx plus a test to run cli installed via package?

File: .github/workflows/release-test.yml
Comment (spacewander) [inline .github/workflows/release-test.yml:41]: There is too much repeated code. Can we reduce them? Or can we just modify the `linux_openresty_common_runner.sh`?

Comment (spacewander) [reply .github/workflows/release-test.yml:41]: Maybe we can add a new step in the build.yml, and remove the `apisix` in git repo, replace `./bin/apisix` with `/usr/bin/apisix`? So thev remain don't need to be changed.

**Comment by Yiyiyimu** [comment]
One problem is which test should we use. Currently we only ignore luarocks check, should we ignore more?

**Comment by Yiyiyimu** [comment]
@spacewander PTAL

**Comment by Yiyiyimu** [comment]
The tarball compressed by Github Actions could be found at
image

File: .github/workflows/build.yml
Comment (Yiyiyimu) [inline .github/workflows/build.yml:65]: need to uncomment this line before merge

File: .github/workflows/build.yml
Comment (Yiyiyimu) [reply .github/workflows/build.yml:61]: Like currently when we run on PR, the last word of `GITHUB_REF` would be `merge`. When it runs when push, it would become the branch name. And when it runs on release branch, branch like `v1.5` would become `1.5`.

File: .github/workflows/build.yml
Comment (Yiyiyimu) [reply .github/workflows/build.yml:74]: Are we changing the release name these days? Yeah it would definitely be of help, so people won't trigger this action when push to their own branch say `verbose`.

And to the last problem, since it would only be triggered when push to release branches, maintainers could control who would trigger this CI.

File: .github/workflows/build.yml
Comment (Yiyiyimu) [reply .github/workflows/build.yml:137]: Cool that could fix the problem. Thanks!

File: .github/workflows/build.yml
Comment (Yiyiyimu) [reply .github/workflows/build.yml:73]: That would be better, thx!

File: Makefile
Comment (Yiyiyimu) [reply Makefile:223]: Thanks! Fixed

File: .github/workflows/build.yml
Comment (spacewander) [reply .github/workflows/build.yml]: We don't need to change the release name. We only need to change the branch name.

Will you submit a final commit to `release/test` branch and show us the final result after changing this? Thanks!

File: .github/workflows/build.yml
Comment (spacewander) [inline .github/workflows/build.yml:134]: Should be `matrix.os_name == 'linux_openresty'`?

Comment (Yiyiyimu) [reply .github/workflows/build.yml]: Fixed. Thanks!

**Review by spacewander — APPROVED** [review]
Will you submit a final commit to release/test branch and show us the final result before we can merge this? Thanks!

**Comment by Yiyiyimu** [comment]
For record unstable test before re-run CI:
```
#   Failed test 'TEST 7: check plugin configuration updating - status code ok'
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 948.
#          got: '404'
#     expected: '200'

#   Failed test 'TEST 7: check plugin configuration updating - response_body - response is expected (repeated req 0, req 0)'
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 1589.
# @@ -1,2 +1 @@
# -passedopentracing
# +fail
# -passedopentracing

#   Failed test 'TEST 7: check plugin configuration updating - grep_error_log_out (req 0)'
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 1149.
# @@ -1,2 +0,0 @@
# -sending a batch logs to 127.0.0.1:5044
# -sending a batch logs to 127.0.0.1:5045
# Looks like you failed 3 tests of 21.
t/plugin/tcp-logger.t ...................... 
Dubious, test returned 3 (wstat 768, 0x300)
```

**Comment by Yiyiyimu** [comment]
Waiting for CI result triggered by `push` on branch `release/test`
https://github.com/apache/apisix/actions/runs/547997620

**Comment by Yiyiyimu** [comment]
@spacewander PTAL

**Comment by starsz** [comment]
LGTM. I think the same things can be done in the dashboard.

**Review by starsz — APPROVED** [review]

**Comment by Yiyiyimu** [comment]
@spacewander Do we need to change the branch titles of previous releases?

**Comment by spacewander** [comment]
Let the previous releases keep frozen.

**Pull request merged**
